lib/session/messageSession.ts

```diff
@@ -0,0 +1,887 @@
+        log.messageSession("[%s] Session with id '%s' is locked until: '%s'.", connectionId,
+          this.sessionId, this.sessionLockedUntilUtc.toISOString());
+        log.error("[%s] Receiver '%s' for sessionId '%s' has established itself.",
+          connectionId, this.name, this.sessionId);
+        log.messageSession("Promise to create the receiver resolved. " +
+          "Created receiver with name: ", this.name);
+        log.messageSession("[%s] Receiver '%s' created with receiver options: %O",
+          connectionId, this.name, options);
+        if (!this._context.messageSessions[this.sessionId!]) {
+          this._context.messageSessions[this.sessionId!] = this;
+        }
+        this._totalAutoLockRenewDuration = Date.now() + (this.maxAutoRenewDurationInSeconds * 1000);
+        await this._ensureTokenRenewal();
+        await this._ensureSessionLockRenewal();
+        await this._resetTimerOnNewMessageReceived();
+      } else {
+        log.error("[%s] The receiver '%s' for sessionId '%s' is open -> %s and is connecting " +
+          "-> %s. Hence not reconnecting.", connectionId, this.name, this.sessionId,
+          this.isOpen(), this.isConnecting);
+      }
+    } catch (err) {
+      this.isConnecting = false;
+      err = translate(err);
+      log.error("[%s] An error occured while creating the receiver '%s': %O",
+        this._context.namespace.connectionId, this.name, err);
+      throw err;
+    }
+  }
+
+  /**
+   * Creates the options that need to be specified while creating an AMQP receiver link.
+   * @ignore
+   */
+  private _createMessageSessionOptions(options: CreateMessageSessionReceiverLinkOptions): ReceiverOptions {
+    const rcvrOptions: ReceiverOptions = {
+      name: this.name,
+      autoaccept: false,
+      // receiveAndDelete -> first(0), peekLock -> second (1)
+      rcv_settle_mode: this.receiveMode === ReceiveMode.receiveAndDelete ? 0 : 1,
+      // receiveAndDelete -> settled (1), peekLock -> unsettled (0)
+      snd_settle_mode: this.receiveMode === ReceiveMode.receiveAndDelete ? 1 : 0,
+      source: {
+        address: this.address,
+        filter: {}
+      },
+      credit_window: 0,
+      onClose: (context) => (options.onClose || this._onAmqpClose)(context).catch(() => { /* */ }),
+      onSessionClose: (context) => (options.onSessionClose || this._onSessionClose)(context).catch(() => { /* */ }),
+      onError: options.onError || this._onAmqpError,
+      onSessionError: options.onSessionError || this._onSessionError,
+      onSettled: options.onSettled || this._onSettled
+    };
+    (rcvrOptions.source as any).filter[Constants.sessionFilterName] = this.sessionId;
+    return rcvrOptions;
+  }
+
+  /**
+   * Ensures that the session lock is renewed before it expires. The lock will not be renewed for
+   * more than the configured totalAutoLockRenewDuration.
+   * @ignore
+   */
+  private async _ensureSessionLockRenewal(): Promise<void> {
```

Comment:
Does a session lock renewal also renews the locks the messages in the session?

Reply:
Yes. As explained in one of the previous comments. Once the link is established then the sdk keeps on renewing the session lock in the background (similar to message lock renewal for message receiver). Renewing the lock for a session takes care of renewing the locks for all the messages in that session.

Reply:
Sweet